vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,743 @@
+      bin_text_read_write_fixed(
+          model_file, reinterpret_cast<char*>(&other_ns_size), sizeof(other_ns_size), "", read, msg, text);
+      for (size_t i = 0; i < other_ns_size; ++i)
+      {
+        namespace_index other_ns;
+        bin_text_read_write_fixed(model_file, (char*)&other_ns, sizeof(other_ns), "", read, msg, text);
+        other_namespaces.insert(other_ns);
+      }
+      exclusions[ns] = other_namespaces;
+    }
+  }
+  else
+  {
+    exclusion_size = exclusions.size();
+    msg << "exclusion_size " << exclusion_size << "\n";
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&exclusion_size), sizeof(exclusion_size), "", read, msg, text);
+    for (const auto& ns_pair : exclusions)
+    {
+      namespace_index ns = ns_pair.first;
+      size_t other_ns_size = ns_pair.second.size();
+      bin_text_read_write_fixed(model_file, (char*)&ns, sizeof(ns), "", read, msg, text);
+      bin_text_read_write_fixed(
+          model_file, reinterpret_cast<char*>(&other_ns_size), sizeof(other_ns_size), "", read, msg, text);
+      for (const auto& other_ns : ns_pair.second)
+      { bin_text_read_write_fixed(model_file, (char*)&other_ns, sizeof(other_ns), "", read, msg, text); }
+    }
+  }
+
+  if (!read || true)
+  {
+    if (!read) msg << "_aml_budget " << budget << "\n";
+    bin_text_read_write_fixed(model_file, reinterpret_cast<char*>(&budget), sizeof(budget), "", read, msg, text);
+  }
+}
+
+// This code is primarily borrowed from expand_quadratics_wildcard_interactions in
+// interactions.cc. It will generate interactions with -q :: and exclude namespaces
+// from the corresponding stride. This function can be swapped out depending on
+// preference of how to generate interactions from a given set of exclusions.
+// Transforms exclusions -> interactions expected by VW.
+interaction_vec quadratic_exclusion_oracle::gen_interactions(const std::map<namespace_index, size_t>& ns_counter,
+    const std::map<namespace_index, std::set<namespace_index>>& exclusions) const
+{
+  std::set<std::vector<namespace_index>> interactions;
+
+  for (auto it = ns_counter.begin(); it != ns_counter.end(); ++it)
+  {
+    auto idx1 = (*it).first;
+    // Only handles ':' for now, extend to handle specific namespaces later as needed
+    if (exclusions.find(idx1) != exclusions.end() && exclusions.at(idx1).find(':') != exclusions.at(idx1).end())
+    { continue; }
+    interactions.insert({idx1, idx1});
+
+    for (auto jt = it; jt != ns_counter.end(); ++jt)
+    {
+      auto idx2 = (*jt).first;
+      // Only handles ':' for now, extend to handle specific namespaces later as needed
+      if (exclusions.find(idx2) != exclusions.end() && exclusions.at(idx2).find(':') != exclusions.at(idx2).end())
+      { continue; }
+      interactions.insert({idx1, idx2});
+      interactions.insert({idx2, idx2});
+    }
+  }
+  interaction_vec v = interaction_vec(interactions.begin(), interactions.end());
```

Comment:
To avoid some copies here you could move the results:
```c++
return interaction_vec(std::make_move_iterator(interactions.begin()), std::make_move_iterator(interactions.end()));
```

Reply:
I think it might be even to here to just pass in the vector